bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java

```diff
@@ -0,0 +1,218 @@
+    }
+
+    public boolean sendPush(@Nullable String recipient, @Nullable String title, @Nullable String message, String type) {
+        boolean result = false;
+
+        logger.debug("sendPush is called for ");
+        logger.debug("Thing {}", thing);
+        logger.debug("Thing Label: '{}'", thing.getLabel());
+
+        PushbulletConfiguration configuration = getConfigAs(PushbulletConfiguration.class);
+        logger.debug("CFG {}", configuration);
+        logger.debug("CFG Name  '{}'", configuration.getName());
+
+        try {
+            logger.debug("Recipient is '{}'", recipient);
+            logger.debug("Title is     '{}'", title);
+            logger.debug("Message is   '{}'", message);
+
+            Push push = new Push();
+            push.setTitle(title);
+            push.setBody(message);
+            push.setType(type);
+
+            if (recipient != null) {
+                if (isValidEmail(recipient)) {
+                    logger.debug("Recipient is an email address");
+                    push.setEmail(recipient);
+                } else if (isValidChannel(recipient)) {
+                    logger.debug("Recipient is a channel tag");
+                    push.setChannel(recipient);
+                } else {
+                    logger.warn("Invalid recipient: {}", recipient);
+                    logger.warn("Message will be broadcast to all user's devices.");
+                }
+            }
+
+            logger.debug("Push: {}", push);
+
+            String request = gson.toJson(push);
+            logger.debug("Packed Request: {}", request);
+
+            Properties headers = new Properties();
+            headers.put(HttpHeader.USER_AGENT, "openHAB / Pushbullet binding " + VERSION.toString());
+            headers.put(HttpHeader.CONTENT_TYPE, MimeTypes.Type.APPLICATION_JSON.asString());
+            headers.put("Access-Token", configuration.getToken());
+
+            logger.debug("Headers: {}", headers);
+
+            try (InputStream stream = new ByteArrayInputStream(request.getBytes(StandardCharsets.UTF_8))) {
+
+                String pushAPI = configuration.getApiUrlBase() + "/" + API_METHOD_PUSHES;
+
+                String responseString = HttpUtil.executeUrl(HttpMethod.POST.asString(), pushAPI, headers, stream,
+                        MimeTypes.Type.APPLICATION_JSON.asString(), TIMEOUT);
+
+                logger.debug("Got Response: {}", responseString);
+                PushResponse response = gson.fromJson(responseString, PushResponse.class);
+
+                logger.debug("Unpacked Response: {}", response);
+
+                stream.close();
+
+                if ((null != response) && (null == response.getPushError())) {
+                    result = true;
+                }
+            }
+            catch (IOException e) {
+                logger.warn("IO problems pushing note: {}", e.getMessage());
+            }
+
+        } catch (java.lang.Error e) {
```

Comment:
Do not catch Error

Reply:
In this case, I have to catch Error to be able to obtain information about the HttpResponseException.

I could check for HttpResponseException, and if that is not the case, rethrow the original error. Would that be o.k. for you?

Reply:
But who Throws an error, please not that errors are different branch from `Exception` it contains things like `OutOfMemmoryError`, `StackOverflow` etc

What exact error do you intend to catch and who threw it?

Reply:
Someone deep inside HttpUtil.executeUrl()... I'll try to provoke an exception so I can provide you with a full stack trace.

In the mean time, I'll check for the HttpResponseException (as mentioned above), and rethrow the exception if it did not contain what I was looking for. Would that be o.k. for you?

Reply:
> Someone deep inside HttpUtil.executeUrl()... I'll try to provoke an exception so I can provide you with a full stack trace.
> 
> In the mean time, I'll check for the HttpResponseException (as mentioned above), and rethrow the exception if it did not contain what I was looking for. Would that be o.k. for you?

Exception would be 'fine' (and are indeed throw from HttpUtil) but you are catching the Error.

Please provide me with it, I would be surprised if its not an Exception but really an Error

Reply:
> Please provide me with it, I would be surprised if its not an `Exception` but really an `Error`

As luck would have it, I am getting "normal" JSON answers to an invalid token just now. Whereas, earlier, it was packaged in the Error. Maybe the API changed since I had written the 1.x action class.

I'll remove the complete error-catching part and wait whether it happens again, then we can talk about how to handle it.

Reply:
> I'll remove the complete error-catching part and wait whether it happens again, then we can talk about how to handle it.

Great, just ping me when you are ready, we should be close to a merge then

Reply:
@martinvw , I did refactor as discussed, and Travis is happy too. I think Jenkins will also be whenever it gets around to build this branch.

Could you give the PushbulletHandler class another look please?